Addresses four compaction service follow on issues.

keith-turner opened the pull request:
While waiting for #1632 to be reviewed  I completed multiple follow on issues for the new compaction service.  I am submitting those as a single PR, but there is a commit for each. I can break them into multiple PRs if desired.

milleruntime commented:
I would like @EdColeman to take a look at the metrics you added for compaction.  I will look over the other commits

milleruntime commented on shell/src/main/java/org/apache/accumulo/shell/commands/CompactCommand.java:
```diff
@@ -147,30 +170,28 @@ public int execute(final String fullCommand, final CommandLine cl, final Shell s
-          "org.apache.accumulo.tserver.compaction.strategies.ConfigurableCompactionStrategy",
-          selectorOpts);
-      compactionConfig.setSelector(selectorCfg);
+    if (cl.hasOption(selectorOpt.getLongOpt())) {
```
If the new options can't be set along with a strategy, shouldn't these two if statements be in an else?

keith-turner replied:
Fixed in abc961a

milleruntime commented on core/src/main/java/org/apache/accumulo/core/conf/Property.java:
```diff
@@ -415,6 +415,10 @@
+  TSERV_COMPACTION_SERVICE_ROOT_THROUGHPUT("tserver.compaction.major.service.root.throughput", "0B",
+      PropertyType.BYTES,
+      "Maximum number of bytes to read or write per second over all major"
+          + " compactions in this compaction service, or 0B for unlimited."),
```
I think a property that is easier to spell is more user friendly.  The last part of the property could be "limit" or "rate.limit".  As a bad speller, I know a word like "throughput" would trip me up.

keith-turner replied:
I like rate.limit better.  Changed to rate.limit in abc961a

milleruntime commented on test/src/main/java/org/apache/accumulo/test/CompactionRateLimitingIT.java:
```diff
@@ -37,50 +40,74 @@
+        long duration = System.currentTimeMillis() - compactionStart;
+        // The rate will be "bursty", try to account for that by taking 80% of the expected rate
+        // (allow
+        // for 20% under the maximum expected duration)
```
Comment got awkward by the formatter.

keith-turner replied:
fixed in abc961a

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/conf/Property.java:
```diff
@@ -415,6 +415,10 @@
   TSERV_COMPACTION_SERVICE_ROOT_PLANNER("tserver.compaction.major.service.root.planner",
       DefaultCompactionPlanner.class.getName(), PropertyType.CLASSNAME,
       "Compaction planner for root tablet service"),
+  TSERV_COMPACTION_SERVICE_ROOT_RATE_LIMIT("tserver.compaction.major.service.root.rate.limit", "0B",
```
I worry that we have added *far* too many new configuration properties for compactions, and that this might create a terrible user experience around compactions. I wonder if some of these could be consolidated into a pluggable configuration strategy object that is less complicated to configure.

ctubbsii replied:
Also, what does "root" refer to here for this property? The description should leave that obvious, rather than require users to refer to the descriptions of other properties that have similar naming schemes in order to guess at what it might mean for this property.

keith-turner replied:
> I worry that we have added far too many new configuration properties for compactions, and that this might create a terrible user experience around compactions.

Could open a new issue and discuss a course of action.  

> Also, what does "root" refer to here for this property? 

This is a compaction service for the root table.

keith-turner replied:
The previous property that shares the same prefix mentions the root table.

ctubbsii replied:
> > I worry that we have added far too many new configuration properties for compactions, and that this might create a terrible user experience around compactions.
> 
> Could open a new issue and discuss a course of action.

For the larger issue, yes. But I'm struggling to keep up with the pace of all these changes, and am raising the point here, to step back and ask the more narrow question of whether these specific properties are needed. If they are, that's fine... we can move on. But, I don't know if I have the mental capacity to circle back and form a coherent discussion around the larger issue that a separate ticket would serve. The time to do that was probably several weeks ago before the wave of new properties got introduced, and that ship may have sailed.

> 
> > Also, what does "root" refer to here for this property?
> 
> This is a compaction service for the root table.

Okay, cool. Thanks. However, I suppose the point of my feedback wasn't to get a response here, but to get a response in the form of an improved description field in the code.

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/compactions/CompactionExecutor.java:
```diff
@@ -94,8 +112,11 @@ public boolean cancel(Status expectedStatus) {
+
       if (canceled && cancelCount.incrementAndGet() % 1024 == 0) {
-        // nMeed to occasionally clean the queue, it could have canceled task with low priority that
+        // need to occasionally clean the queue, it could have canceled task with low priority that
         // hang around. Avoid cleaning it every time something is canceled as that could be
         // expensive.
```
This sentence could be cleaned up to make it more understandable. What does "it" refer to? And does it mean it could "contain a canceled task", or that it could "have canceled a task"?

keith-turner replied:
it refers to the queue.. I can clean this up

keith-turner replied:
Changed comment in 79611ee

ctubbsii replied:
Awesome, thanks! A minor cleanup would be great!

EdColeman commented:
Looking at the metrics generated by CompactionExecutorsMetrics.  

My initial concern was that there could be cardinality issues if the metrics names are unbounded - that does not seem to be an issue.  Using the current code in your branch, I quickly added a delay in the FateConcurrencyIT test that forces a pause while a compaction FATE transaction is running so that the metrics system has time to write records to the test output file.  The generated record for the IT test tserver looks like:

`1596645249079 tserver.compactionExecutors: Context=tserver, ProcessName=TabletServer, Hostname=ip-xxx-xxx-xxx-xxx, default_large_queued=0, default_large_running=0, default_medium_queued=0, default_medium_running=0, default_small_queued=0, default_small_running=0, meta_huge_queued=0, meta_huge_running=0, meta_small_queued=0, meta_small_running=0, root_small_queued=0, root_small_running=0, root_huge_queued=0, root_huge_running=0, default_any_queued=0, default_any_running=1`

This looks good, but why is default_any_running=1 but nothing in default large, medium or small?  Is default_any a separate queue, or should it be the sum of all default_any_ ?

With this forced test, I know the FATE transaction is "IN_PROGRESS"  - it uses a slow compaction runner that pauses 250ms in the iterator before each next() call - could that be changing the "running" to something where the thread is paused and that does not count as "running" 

Before digging I just want to validate that my assumptions w.r.t the names is accurate.

EdColeman commented on test/src/main/java/org/apache/accumulo/test/CompactionRateLimitingIT.java:
```diff
@@ -37,50 +40,73 @@
+        // (allow for 20% under the maximum expected duration)
+        assertTrue(String.format(
+            "Expected a compaction rate of no more than %,d bytes/sec, but saw a rate of %,f bytes/sec",
+            (int) 0.8d * RATE, 1000.0 * bytesWritten / duration),
```
```suggestion
            (int) (0.8d * RATE), 1000.0 * bytesWritten / duration),
```

keith-turner commented:
> This looks good, but why is default_any_running=1 but nothing in default large, medium or small? Is default_any a separate queue, or should it be the sum of all default_any_ ?

@EdColeman do you still have the source for the test that emitted these metrics?  I would like to see the compaction config used in the test.

EdColeman commented:
Not source code as such - the configuration would have been the defaults.

What I do is hack one of the IT tests that use the slow compaction runner and then just insert a hard Thread.sleep(xxx) - I usually let the test fail because of timeout, so I have as much time as possible for things to report.  

One candidate would be in org.apache.accumulo.test.functional.FateConcurrencyIT.multipleCompactions() at line 496 - right before the compactions are cancelled.  The test starts 4 compactions with a slow iterator so there's plenty of time to have metrics reported.

When the test is running, you can use jconsole to examine jmx - or use the metric file output - with this the should be multiple metrics measurements.  You may need to adjust the metrics config so that the metrics files are written - they will end up in the target directory of the ITs.

keith-turner commented:
@EdColeman @milleruntime @ctubbsii I would like to merge this before it ends up w/ conflicts.  I think I addressed all comments.  Did I miss anything or where there any other questions?

milleruntime reviewed (approved):
I think so, I didn't look at the Metrics stuff though since Ed was.

EdColeman commented:
From what I had looked at, you answered the concerns that I had.  It seemed like you were going to try to manually examine the metrics emitted, do you have a chance?  More out of curiosity, I'm not suggesting that it needs to hold this up.

keith-turner commented:
Since this branch has been sitting for a while, locally I merged this branch to main and reran all ITs matching *Compaction*IT*.  The test were all a success.
